### PR TITLE
Add FixedSizeWindowFunction interface

### PR DESCRIPTION
**Pull request by miniway**
It would let a special Window function limit output records

### CODE REVIEW

File: presto-spi/src/main/java/io/prestosql/spi/function/LimitWindowFunction.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.function;
+
+public interface LimitWindowFunction
```
Comment (xerial) [inline presto-spi/src/main/java/io/prestosql/spi/function/LimitWindowFunction.java]: I think using FixedLengthWindowFunction is more straightforward to understand the role of this interface. 

For `Limit(ed)`, it's not sure what is limited.

File: presto-spi/src/main/java/io/prestosql/spi/function/LimitWindowFunction.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+{
+    int[] select(int positionCount);
```
Comment (xerial) [inline presto-spi/src/main/java/io/prestosql/spi/function/LimitWindowFunction.java]: This method needs a description. And also, the selection doesn't happen when this method is called, the right naming would be returning `int[] selectedPositions` if I understand correctly.

File: presto-main/src/main/java/io/prestosql/operator/WindowOperator.java [hunk 3]
```diff
@@ -607,6 +623,68 @@ public TransformationState<Page> process(WindowPartition partition)
         }
     }
 
+    private class LimitWindowPartitionsToOutputPages
```
Comment (xerial) [inline presto-main/src/main/java/io/prestosql/operator/WindowOperator.java]: FixedSizeWindowPartitionsToOutputPages

File: presto-main/src/main/java/io/prestosql/operator/WindowOperator.java [hunk 4]
```diff
@@ -303,6 +312,13 @@ public WindowOperator(
                     .flatMap(this::pagesIndexToWindowPartitions)
                     .transform(new WindowPartitionsToOutputPages());
         }
+        else if (limitWindowFunction.isPresent()) {
```
Comment (xerial) [inline presto-main/src/main/java/io/prestosql/operator/WindowOperator.java]: With this condition, it looks like if the disk spill is enabled, LimitWindowPartitionToOutputPages will never be used.

File: presto-main/src/main/java/io/prestosql/operator/WindowOperator.java [hunk 5]
```diff
@@ -607,6 +623,68 @@ public TransformationState<Page> process(WindowPartition partition)
...
+            }
+
+            // process whole partition
+            // TODO: When a single partition is too large, OOM could happen
+            while (partition.hasNext()) {
+                partition.processNextRow(tempPageBuilder);
+            }
+            int[] positions = function.select(tempPageBuilder.getPositionCount());
+            copyPositions(tempPageBuilder, pageBuilder, positions);
+
```
Comment (xerial) [inline presto-main/src/main/java/io/prestosql/operator/WindowOperator.java]: Yeah. The memory consumption of tempPageBuilder looks huge for large input tables. Can't we have a reservoir buffer within this transformation class and perform the sampling here?

File: presto-main/src/main/java/io/prestosql/operator/WindowOperator.java (hunk 4 above)
Comment (miniway) [reply presto-main/src/main/java/io/prestosql/operator/WindowOperator.java]: it’s an intentional as we would not use disk spill in the near futur

File: presto-main/src/main/java/io/prestosql/operator/WindowOperator.java (hunk 5 above)
Comment (xerial) [reply presto-main/src/main/java/io/prestosql/operator/WindowOperator.java]: I guess the memory usage difference is:
- current: (WindowPartition size) x (the number of unique users)
- optimum: (reservoir size) x (the number of unique users)

Or tempPageBuilder can be quickly reduced to the reservoir size as Presto processes WindowPartition?

Comment (miniway) [reply presto-main/src/main/java/io/prestosql/operator/WindowOperator.java]: Not sure how often we have a partition-size >>> sample-size. 
One idea is a cascading page buffer. For example, tempBuffer1.ifFull -> select().create(tempBuffer2), tempBuffer2.ifFull -> select().create(tempBuffer3) an so on.

Comment (xerial) [reply presto-main/src/main/java/io/prestosql/operator/WindowOperator.java]: As we only need to maintain a fixed-size reservoir and the input records up to its size, adding a sampling around partition.processNextRow(...) line looks more natural to me. Isn't it possible?

File: presto-spi/src/main/java/io/prestosql/spi/function/FixedSizeWindowFunction.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.function;
+
+public interface FixedSizeWindowFunction
+{
+    int[] selectPositions(int positionCount);
```
Comment (xerial) [inline presto-spi/src/main/java/io/prestosql/spi/function/FixedSizeWindowFunction.java:18]: selectedPosition?

Comment (miniway) [reply presto-spi/src/main/java/io/prestosql/spi/function/FixedSizeWindowFunction.java:18]: well, it selects positions by calling this. caller might not know whether it selects at this point or returns selected-positions.

Comment (xerial) [reply presto-spi/src/main/java/io/prestosql/spi/function/FixedSizeWindowFunction.java:18]: OK. Understood

File: presto-main/src/main/java/io/prestosql/operator/WindowOperator.java [hunk 7]
```diff
@@ -607,6 +623,70 @@ public TransformationState<Page> process(WindowPartition partition)
...
+                    copyPositions(partition, positions);
+                }
+
+                // Output the remaining page if we have anything buffered
+                Page page = pageBuilder.build();
+                pageBuilder.reset();
+                return TransformationState.ofResult(page, false);
+            }
+
+            // process whole partition without output channels for unwanted copy and memory usage
```
Comment (xerial) [inline presto-main/src/main/java/io/prestosql/operator/WindowOperator.java:663]: 👍

File: presto-main/src/main/java/io/prestosql/operator/WindowOperator.java [hunk 8]
```diff
@@ -607,6 +623,70 @@ public TransformationState<Page> process(WindowPartition partition)
...
+            while (partition.hasNext()) {
+                partition.processFixedSizeWindowNextRow(framedFunction, windowBlockBuilder);
+            }
+            int[] positions = function.selectPositions(windowBlockBuilder.getPositionCount());
+            copyPositions(partition, positions);
```
Comment (xerial) [inline presto-main/src/main/java/io/prestosql/operator/WindowOperator.java:668]: ok. It looks good. So this part is 
- Update the reservoir buffer with processFixedSizeWindowNextRow
- Get the selected positions
- Copy the rows at the selected positions

**Review by xerial — APPROVED** [review]
LGTM now!

**Pull request merged**